src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,137 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int INITIAL_CAPACITY = 10;
+    private static final int INITIAL_SIZE = 0;
+    private static final int INITIAL_INDEX = 0;
+    private static final String TEXT_OF_EXCEPTION = "Invalid index outside"
+            + " the dimension of the array";
+    private int size;
+    private T[] elementData;
+
+    public ArrayList() {
+        size = INITIAL_SIZE;
+        elementData = (T[]) new Object[INITIAL_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            increaseElementDataLength();
+        }
+        addValueToElementData(value);
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < INITIAL_INDEX) {
+            throw new ArrayListIndexOutOfBoundsException(TEXT_OF_EXCEPTION);
+        } else if (index == elementData.length - 1) {
+            increaseElementDataLength();
+            addValueToElementData(value);
+        } else {
+            rightShiftOfElements(value, index);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        T[] bufferArray = (T[]) new Object[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            bufferArray[i] = list.get(i);
+        }
+        if (elementData.length - size < list.size()) {
+            increaseElementDataLength();
+        }
+        System.arraycopy(bufferArray, INITIAL_INDEX, elementData, size, bufferArray.length);
+        size += bufferArray.length;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < INITIAL_INDEX) {
+            throw new ArrayListIndexOutOfBoundsException(TEXT_OF_EXCEPTION);
+        } else {
+            return elementData[index];
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < INITIAL_INDEX) {
+            throw new ArrayListIndexOutOfBoundsException(TEXT_OF_EXCEPTION);
+        } else {
+            elementData[index] = value;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < INITIAL_INDEX) {
+            throw new ArrayListIndexOutOfBoundsException(TEXT_OF_EXCEPTION);
+        }
+        if (index == size - 1) {
+            size--;
+            return elementData[index];
+        } else {
+            T removeValue = elementData[index];
+            leftShiftOfElements(index);
+            return removeValue;
+        }
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (element == null) {
+                if (null == elementData[i]) {
```

Comment:
use the next approach to check elements equality, and just call existing `remove(i)` method
```
 if ((elementData[i] == element  || elementData[i] != null && elementData[i].equals(element))) {
                return remove(i);
            }
 ```